### PR TITLE
BUGFIX: add 'validationErrorMessage' as a supported option

### PR DESCRIPTION
**Pull request by MiauzGenau**
Add 'validationErrorMessage' as an supported option, for the RegularExpressionValidator.

resolved: https://github.com/neos/neos-ui/issues/3691

**Checklist**
- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

### CODE REVIEW

**Review by kdambekalns — CHANGES_REQUESTED** [review]
And what about actually using it, too?

**Comment by fcool** [comment]
If the aim is to change the error message, there are plenty of possibilities. You could use the code and provide an own translation. Or you could override the translation flow provides. Why would you have the message changeable for each validator instance?
And given the use case is convincing: Why only with the regex validator? And not with all validators?

Are you sure, this is a bugfix and not a feature?

**Comment by kdambekalns** [comment]
> If the aim is to change the error message, there are plenty of possibilities.

Obviosuly the option was added to the UI in https://github.com/neos/neos-ui/pull/2996 but that (now?) breaks with Flow.

As to "why only here", then probably because most other default messages are good enough, but "does not match pattern" holds next to no useful information. To which I can agree. 🙃

**Comment by fcool** [comment]
Ah... and that's why it (the new message) is not used in Flow itself.

But that feels like a bug in Neos, doesn't it? Flow shouldn't allow a property to be changed, because of Neos setting it "accidently", to fire own logic? Maybe we should add an additional validator in Neos then?

**Comment by kdambekalns** [comment]
> But that feels like a bug in Neos, doesn't it? Flow shouldn't allow a property to be changed, because of Neos setting it "accidently", to fire own logic? Maybe we should add an additional validator in Neos then?

- Well, I can only assume it used to work–and now breaks due to Flow being more strict? 🤷‍♂️
- If this fixes a bug, but introduces weird behaviour (the message being "supported", but not used), I am against it.
- An additional validator feels weird, given the usecase is valid for (pure) Flow applications, too.

**Comment by MiauzGenau** [comment]
I created a new PR: https://github.com/neos/flow-development-collection/pull/3141
As a FEATURE draft.
For now I only did it for the RegularExpressionValidator, if this looks good/ better to you I would also adjust the other validators and close this PR.

**Comment by mhsdesign** [comment]
> But that feels like a bug in Neos, doesn't it? Flow shouldn't allow a property to be changed, because of Neos setting it "accidently", to fire own logic? Maybe we should add an additional validator in Neos then?

yeeeees true. The current situation is suboptimal and flow should not suffer from this. Agreed ^^.
We are currently evaluating better concepts than using flows validators for node properties, so they can be validated in the cr: https://github.com/neos/neos-development-collection/issues/4631

The "auto fire" was introduced via https://github.com/neos/neos-ui/pull/2351 and actually caused much headache, as the javascript validators and phpvalidators are not always the same and validate differently. This new message feature is just an example.

As this feels indeed quite hacky on php land

> If the aim is to change the error message, there are plenty of possibilities. You could use the code and provide an own translation. Or you could override the translation flow provides. Why would you have the message changeable for each validator instance?

i would say the triggering logic in the neos ui php code should be adjusted. As this neos ui property server validation is already super hacky (see https://github.com/neos/neos-ui/pull/2351) we should fix it there (as bugfix) and keep all the ugly code at one place and when the time comes remove it.

TLTR: I would like to close this in favour of fixing it in the Ui

**Review by kitsunet — CHANGES_REQUESTED** [review]
Pre-emptive block as we definitely should not merge this, if anything the feature PR linked in here or fix it in the Neos UI (probabably better). This change would introduce an option that has absoluitely no use or result in Flow and will be very misleading.

**Comment by lorenzulrich** [comment]
@mhsdesign How would you fix this in the UI? It seems that all validator options are passed to the PHP validator, regardless of whether they are "UI-only" or also valid for the PHP validation.

Wouldn't it make sense to distinguish between the kind of option?

```
'Vendor.Site:Content.YourNodeTypeName':
  properties:
    lastName:
      type: string
      ui:
		label: 'Last Name'
        inspector:
          group: 'contactData'
      validation:
        'Neos.Neos/Validation/RegularExpressionValidator':
          regularExpression: '...'
          neosUi:
            validationErrorMessage: 'You must set a last name.'
```

**Comment by mhsdesign** [comment]
@lorenzulrich i moved the discussion and my answer over here: https://github.com/neos/neos-ui/issues/3691#issuecomment-1898100119

**Pull request closed without merging**
